app/controllers/matches_controller.rb

```diff
@@ -13,22 +13,38 @@ def show
   # PATCH/PUT /matches/1
   def update
     new_state = match_params['state']
-    if @match.update(match_params)
-      if new_state == 'finished'
-        result = PopulateMatchBelowAndSave.call(match: @match) unless @match.group_match?
-        unless result.success?
-          render json: { error: 'Moving Team one stage down failed' }, status: :unprocessable_entity
-          return
-        end
+
+    Match.transaction do
+      if @match.update(match_params)
+        return handle_match_end if new_state == 'finished'
+
+        render json: @match
+      else
+        render json: @match.errors, status: :unprocessable_entity
+        raise ActiveRecord::Rollback
       end
-      render json: @match
-    else
-      render json: @match.errors, status: :unprocessable_entity
     end
   end
 
   private
 
+  def handle_match_end
+    return if @match.group_match?
+
+    if @match.winner.nil?
+      error_message = 'Stopping undecided Matches isn\'t allowed in playoff stage'
+      render json: { error: error_message }, status: :unprocessable_entity
+      raise ActiveRecord::Rollback
+    end
+
+    result = PopulateMatchBelowAndSave.call(match: @match)
+    return if result.success?
```

Comment:
i think a `unless result.success?` is easier to read here

Reply:
if != unless :thinking:

Reply:
i'm aware. but i think a big unless block until the end of the method is easier to read